### DISCUSSION TITLE
Best way to get the current root node?

### DISCUSSION THREAD

**Discussion by matthewKeville** [nvim-tree/nvim-tree.lua#2549] [Q&A] [answered]
In my work flow I would like to be able to use Telescope from the root of the current tree (I keep different trees & don't sync with cwd). This requires me be able to determine the path of that node.

If I needed to Telescope into a child of the root this isn't that difficult using `nvim-tree.api.get_node_under_cursor`.

When it comes to using the root, the solution I came up with seems hacky and I am curious if there is a better way to find
the root.

```lua
local api = require("nvim-tree.api")


--- @return string|nil
local function root_node_dir_path()

  local node = api.tree.get_node_under_cursor()
  if not node then
    return
  end

  while ( node.parent ) do
    node = node.parent
  end

  return node.absolute_path

end

```
I can use this approach whenever my current tree line is not at the root

But if it is, i.e. at '~/dotfiles/..` then I can't use this method because it appears that the root isn't a node.

```md
~/dotfiles/..
> [] .git
> [] _config
> [] desktop
> [] scripts
   .gitignore
   link.sh
   read.me
```
Is there a better way to find this that doesn't rely on the cursor position?
Why isn't there something in the API to expose the root? ( Or is there..? )

**Comment by alex-courtis** [comment]
I use something surprisingly similar for telescope: https://github.com/alex-courtis/arch/blob/9e7e3e8d7aeabe9e49bc7c52eb128bdf21f3d9e5/config/nvim/lua/amc/plugins/nvt.lua#L43

I do, however, sync global cwd everywhere.

We've recently added some new API. You could do something like this to retrieve the local working directory of the tree.
```
  local api = require("nvim-tree.api")
  local log = require("nvim-tree.log")

  local winid = api.tree.winid()
  log.line("dev", "winid=%s", vim.inspect(winid))
  local cwd = vim.fn.getcwd(winid)
  log.line("dev", "cwd=%s", vim.inspect(cwd))
```

**Reply by alex-courtis** [reply, marked as answer]
Try `:help nvim-tree-api.tree.get_nodes()`

The root will contain `absolute_path`

**Reply by matthewKeville** [reply]
I may or may not have drew inspiration from that https://github.com/alex-courtis/arch/blob/9e7e3e8d7aeabe9e49bc7c52eb128bdf21f3d9e5/config/nvim/lua/amc/plugins/nvt.lua#L43

**Reply by alex-courtis** [reply]
More complete implementation in the wiki: https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#find-and-focus-directory-with-telescope
